### PR TITLE
8285470: Improve handling of @inheritDoc

### PR DESCRIPTION
**Pull request by pavelrappo**
The only taglet that along with its own tag needs to know the immediately enclosing tag is `InheritDocTaglet`:

    @return {@inheritDoc}
    @throws NullPointerException {@inheritDoc}
    @param p {@inheritDoc}

However, the immediately enclosing tag is unconditionally passed to all taglets. If we stop passing it and make `InheritDocTaglet` compute it instead, the code becomes cleaner.

While reviewing, particularly note these benefits of the proposed change:

 * taglet-handling code knows less about `@inheritDoc`, and
 * `InheritDocTaglet` receives its own tag, not the tag that encloses it

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 reviews required, with at least 1 reviewer)

### Issue
 * [JDK-8285470](https://bugs.openjdk.java.net/browse/JDK-8285470): Improve handling of @inheritDoc

### Reviewers
 * [Jonathan Gibbons](https://openjdk.java.net/census#jjg) (@jonathan-gibbons - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8361/head:pull/8361` \
`$ git checkout pull/8361`

Update a local copy of the PR: \
`$ git checkout pull/8361` \
`$ git pull https://git.openjdk.java.net/jdk pull/8361/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8361`

View PR using the GUI difftool: \
`$ git pr show -t 8361`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8361.diff">https://git.openjdk.java.net/jdk/pull/8361.diff</a>

</details>

### CODE REVIEW

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/InheritDocTaglet.java [hunk 1]
```diff
@@ -56,39 +56,34 @@ public InheritDocTaglet() {
...
     /**
-     * Given an element, a {@code DocTree} in the element's doc comment
-     * replace all occurrences of {@code {@inheritDoc}} with documentation from its
-     * superclass or superinterface.
+     * Given an element and {@code @inheritDoc} tag in that element's doc comment,
+     * returns the (recursive) expansion of that tag.
+     *
+     * This method does not expand all {@code {@inheritDoc}} tags in the given
+     * element's doc comment. To do this, the method must be called for every
+     * such tag.
```
Comment (ExE-Boss) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/InheritDocTaglet.java:64]: This should probably be wrapped in a `<p>` tag:
```suggestion
     * <p>This method does not expand all {@code {@inheritDoc}} tags in the given
     * element's doc comment. To do this, the method must be called for every
     * such tag.</p>
```

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java [hunk 2]
```diff
@@ -1412,60 +1412,51 @@ boolean ignoreNonInlineTag(DocTree dtree) {
...
                                         List<? extends DocTree> tags,
                                         boolean isFirstSentence)
     {
-        return commentTagsToContent(holderTag, element, tags, isFirstSentence, false);
+        return commentTagsToContent(element, tags, isFirstSentence, false);
     }
 
     /**
-     * Converts inline tags and text to text strings, expanding the
+     * Converts inline tags and text to Content, expanding the
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java:1428]: either `content` or `{@code Content}`

**Review by jonathan-gibbons — COMMENTED** [review]
This one is somewhat questionable.  

Generally, the trees in `javac` and `javadoc` do not store parent pointers (to save space), leading to a general coding philosophy of "collect any necessary information as you walk down the tree".  I understand why you want to get rid of the extra parameters, but it is disappointing to see the code use a tree-walk (even if we have recently improved it) to re-acquire info that was already available a few stack-frames up.

Since the magic/use is primarily in `TagletWriter/Impl` with most other edits being to propagate the info, is there any better way to make the `holder` available, without having to call `getDocTreePath`? For example, `HtmlDocWriter` creates a `TagletWriter round about line 381: is there a way we could stash the holder in that object for later retrieval by the `InheritDocTaglet`.

I would hate for this PR to be seen as a precedent for more widespread use of `get[Doc]TreePath`.

**Comment by pavelrappo** [comment]
I think that this improvement in readability wins over the miniscule performance hit it brings. Currently, walking down the stack feels like telling a story with too many details too soon.

Speaking of irrelevant details. The related `DocFinder` subsystem is even worse. Its `Input` and `Output` classes are plagued with fields for specific taglets. I'm not a fan of this "a bag of everything" pattern.

**Comment by pavelrappo** [comment]
> This one is somewhat questionable.
> 
> Generally, the trees in `javac` and `javadoc` do not store parent pointers (to save space), leading to a general coding philosophy of "collect any necessary information as you walk down the tree". I understand why you want to get rid of the extra parameters, but it is disappointing to see the code use a tree-walk (even if we have recently improved it) to re-acquire info that was already available a few stack-frames up.
> 
> Since the magic/use is primarily in `TagletWriter/Impl` with most other edits being to propagate the info, is there any better way to make the `holder` available, without having to call `getDocTreePath`? For example, `HtmlDocWriter` creates a `TagletWriter round about line 381: is there a way we could stash the holder in that object for later retrieval by the `InheritDocTaglet`.
> 
> I would hate for this PR to be seen as a precedent for more widespread use of `get[Doc]TreePath`.

I think that this improvement in readability wins over the miniscule performance hit it brings. Currently, walking down the stack feels like telling a story with too many details too soon.

Speaking of irrelevant details. The related `DocFinder` subsystem is even worse. Its `Input` and `Output` classes are full of fields used by specific taglets. I'm not a fan of this "a bag of everything" pattern.

**Comment by pavelrappo** [comment]
Thanks, Jon.

For the reader's convenience, let me link to that tree-search support improvement that you mentioned: JDK-8267690 (https://github.com/openjdk/jdk/pull/8369)

**Pull request closed without merging**